DS-2862 Fix legacy embargo checks on DSpace 6

minurmin opened the pull request:
https://jira.duraspace.org/browse/DS-2862

While running the legacy embargo lifter it turned out that the conditions on legacy embargo checker to detect anonymous read rights (i.e. "CHECK WARNING: Item ..., Bundle ORIGINAL allows READ by Group Anonymous) were not working correctly, producing NullPointerExceptions (trying to get name from null rp.getGroup()). This code fixes the conditions to show the warnings, assuming that the "timed" resource policy is correct compared to the value set in dc.date.embargo field.

The code is tested with data migrated from a legacy DSpace 1.7.3 installation. Initially "dspace migrate-embargo" was not run (which it should have), which was the reason legacy embargo appeared to be broken in DSpace 6 - and Monika Mevenkamp's fix to call legacy embargo lifter on embargo manager (=commented in DSpace 3) appeared to fix this functionality.

minurmin committed:
DS-2862 Activate legacy embargo lifter, fix conditions on embargo check warnings

terrywbrady commented:
@minurmin , the code changes in this PR look very straight forward, but more information is needed to reproduce the error and to verify the fix.  Could you update the PR description to describe the sequence of actions that would be needed to reproduce the original error?

In order to verify this bug, is it necessary to create the embargo in a DSpace 1.7/3.x version of DSpace, or can the bug be reproduced in DSpace 6.x?

I added the "needs discussion" label on this PR until more details are added.

minurmin commented:
The NullPointerException is reproducible in 6.x by adding a embargo.field.lift field (e.g. dc.date.embargo with a date in the future) to any item - but keep the policies unchanged (i.e. readable by ANONYMOUS). When running the embargo-lifter, it produces error message like:

`Failed attempting to lift embargo, item=123456789/1234: java.lang.NullPointerException`

After applying the fix the message will be like:

```
CHECK WARNING: Item 123456789/1234, Bundle ORIGINAL allows READ by Group Anonymous
CHECK WARNING: Item 123456789/1234, Bitstream doc.pdf (in Bundle ORIGINAL) allows READ by Group Anonymous
```

I tested this again with 6.1-based install and it seems that NullPointerExceptions may pop up with both `dspace migrate-embargo` and `dspace embargo-lifter` commands if there are Anonymous permissions.

However, additional discussion might be needed on why legacy embargo lifter line was commented out in DSpace 3 in the first place. Are the known cases (other than applying legacy+new embargo simultaneously) where the embargo lifter might mess the policies? I'm referring to message in dspace-tech https://groups.google.com/d/topic/dspace-tech/ST0kJrrD-MI/discussion "I'm guessing it was removed in 4.2 to prevent ALL custom policies from being wiped out inadvertently". If this is the case, the effect could be compensated with a new config parameter like "enable-legacy-embargo".

terrywbrady commented:
### Testing notes

1. Add the following to local.cfg

    embargo.field.terms = dc.embargo.terms
    embargo.field.lift = dc.embargo.lift

2. Add the same fields to the metadatafieldregistry
3. Add the following to input-forms.xml

       <field>
         <dc-schema>dc</dc-schema>
         <dc-element>embargo</dc-element>
         <dc-qualifier>terms</dc-qualifier>
         <repeatable>false</repeatable>
         <label>Embargo</label>
         <input-type>onebox</input-type>
         <hint>Date or "forever"</hint>
         <required></required>
       </field>

4. Submit an item with a past embargo date and an item with a future embargo date
5. Look at the full item metadata produced by the embargo setter
6. Look at the bistream policies created on the new bitstreams
7. Run the embargo lifter
8. Re-look at the full item metadata produced by the embargo setter
9. Re-look at the bistream policies created on the new bitstreams

terrywbrady reviewed (changes requested):
@minurmin , thanks for submitting this change.  This functionality in DSpace is not well understood, and I think you have highlighted some issues that make it more confusing.

Your change to DefaultEmbargoSetter does resolve the Null Pointer Exception.  I agree that change should be incorporated into the code.

Looking at both the metadata and the bitstream policies, I do not think that the call to lifter.liftEmbargo should be removed.  I recommend adding a comment next to that line indicating that the "lift" process is performed through the dates on the authorization process.

As you described, there is a misleading warning when a record with a future policy date is found.  I think the test that generates the following error should examine the policy dates before reporting the error.

CHECK WARNING: Item 123456789/25, Bundle ORIGINAL allows READ by Group Anonymous
CHECK WARNING: Item 123456789/25, Bitstream IMG_5032.jpg.jpg (in Bundle ORIGINAL) allows READ by Group Anonymous

minurmin commented:
@terrywbrady , to make sure I understood correctly, are you suggesting the following changes:

1) If the item has an Anonymous resource policy with a start date in the future the lifter.liftEmbargo call should not made at all (=handled by new embargo functionality). This way there would be clear precedence that new embargo functionality (if used), "overrides" the old one.

2) Add policy date check as in 1) to DefaultEmbargoSetter to prevent unnecessary check warnings.

3) Add comment like "By default, the lift process is performed through the dates on the authorization process. Alternatively the date specified in item metadata is used" before lifter.liftEmbargo call.

terrywbrady commented:
@minurmin , under what conditions is a call to liftEmbargo still needed?  I believe that all legacy embargo settings should be migrated with the embargo migration tool.  If that is true, the only task that the embargo lifter would need to perform is to remove the metadata element.  Let me know if my assumptions are not correct.

terrywbrady commented:
@minurmin , could you reply to my question on this PR?

minurmin commented:
The PR was made under the assumption that the legacy embargo would be used in DSpace 6 (I wasn't quite clear whether this should be supported or not). But indeed, perhaps it is simpler to concentrate just fixing the embargo migration tool so my question above can be discarded.
I'll run additional tests using legacy data with policy date check added and update the PR accordingly.

tdonohue commented:
Just a quick note, in case it's unclear, legacy embargo *should still work in 6.x*.  We have a ticket open to possibly remove it from 7.x: https://jira.duraspace.org/browse/DS-2588  However, that effort requires ensuring the new embargo system can be improved to meet all the use cases of the legacy embargo.  Currently, there are use case gaps (e.g. I don't believe a new embargo can be set via SWORD and maybe not via REST either...but a legacy embargo can be set via any import mechanism, as it's just a metadata field).

tdonohue commented on dspace-api/src/main/java/org/dspace/embargo/EmbargoServiceImpl.java:
```diff
@@ -151,7 +151,7 @@ public void liftEmbargo(Context context, Item item)
     {
         // new version of Embargo policies remain in place.
-        //lifter.liftEmbargo(context, item);
+        lifter.liftEmbargo(context, item);
```
This change looks to be unnecessary. This line of code was explicitly commented out when the new embargo system was added in DSpace 3.  The comment above (while badly worded) seems to note that traditional embargos get "translated" into new embargo system (as resource policy dates), so this explicit `liftEmbargo` shouldn't be necessary.

Here's the DSpace 3 commit which added new embargo (and commented out this line): https://github.com/DSpace/DSpace/commit/4b9b206d8c1455cc556688c726ae0ed596d25205#diff-b11c7ae8084a53137fcf0dca05b9fce6R179

All that said, if you are finding that traditional embargos are *not getting lifted properly in 6.x*, then that is definitely a bug.  But, since this line has been commented out since 3.x, I don't think it's the source of the bug (as traditional embargoes work fine in 5.x).

terrywbrady replied:
My recommendation is that the embargo lifter should only remove the lift date metadata.  It should not attempt to modify the access policies for bitstreams.

minurmin replied:
I have to admit I'm still a bit confused about this one. 

In our test server, legacy embargo was not working in DSpace 6 (at least with our data). By legacy embargo, I mean running the daily cron script dspace embargo-lifter (and I definitely assumed that lifting the embargo involves modifying the access policies for bitstreams). At this stage migrate-embargo utility was not used.

Then I found DS-2862 where lifter.liftEmbargo line was added (again) in DSpace 5. Based on this I assumed that traditional embargoes do not work correctly in 5.x either. But: based on @tdonohue :s comment I believe that there is a bug with lifting traditional embargoes. If the line  lifter.liftEmbargo is supposed to be commented, I suggest that the changes introduced in DSpace 3 commit are investigated w.r.t DSpace 6 code to fix the legacy embargo properly.

Other changes in this PR so far should not affect the lifting functionality itself, just reporting about reporting about Anonymous access to bitstreams (and I agree that this should yet be supplemented with policy date check).

tdonohue replied:
Hi @minurmin : It's my understanding that the `lifter.liftEmbargo()` line is no longer needed because traditional embargoes essentially get "translated" into new embargoes automatically by the `setEmbargo()` method. 

https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/embargo/DefaultEmbargoSetter.java#L88-L108

Notice how that method calls `getEmbargoTermsAsDate()`, which reads the embargo date set in embargo metadata fields:
https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/embargo/EmbargoServiceImpl.java#L113-L146

And, then `setEmbargo()` takes that date and creates Resource Policies on all bundles/bitstreams to essentially "transform" this into a new embargo.

That's why we believe the `liftEmbargo()` method should be unnecessary.  Essentially, a "traditional embargo" should be auto-translated into a "new embargo" at the time the item is saved into the archive, as `setEmbargo()` is called in the `InstallItemService.finishItem()` method which installs the item into the Archive: https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/content/InstallItemServiceImpl.java#L236

So, if this isn't working properly, we may want to more closely walk through / debug the code here. Based on the logic I see, `liftEmbargo()` should be unnecessary (and should remain commented out).  But, that may mean there's a bug somewhere else that we are missing.

minurmin replied:
Ok, I think I finally got it. Many thanks for the clarification, it seems that I did not quite understand the meaning of "legacy embargo" in DSpace 6 context. 

Confirmed: if the item is submitted to DSpace 6 with an editable dc.date.embargo metadata field, default embargo setter sets the timed policy as expected.

However, if the database is migrated from a pre-3.0 DSpace, the timed policies are _not_ in place (i.e. an item with legacy embargo has no read policy on the bitstream). One has to run dspace migrate-embargo -a to generate the policies - which I didn't do since I assumed (based on description of DS-2588, e.g. "This type of embargo can only be lifted by running the ./dspace embargo-lifter command") that the legacy embargo would work in 6.0 as in 1.7 (=i.e. legacy embargo would involve no timed policies). So the "fix" would involve updating the documentation to stress out the fact that it is necessary to run  migrate-embargo if updating from pre 3.0 DSpace, and note that the intended behaviour of the legacy embargo lifter utility has been changed. At least:
* https://wiki.duraspace.org/display/DSDOC6x/Embargo#Embargo-Pre-3.0EmbargoMigrationRoutine (contains no explanation when the migration utility should be run)
* https://wiki.duraspace.org/display/DSDOC6x/Upgrading+DSpace (makes no mention about migrating embargoes)

Considering the immediate problem of getting embargoes to work with DSpace 6 I now agree that the line lifter.liftEmbargo is not needed after all. The basic functionality more or less works as expected, provided that one understands the intended usage. 

It should be noted that with post 3.0 code customized embargo lifters (e.g. https://groups.google.com/d/topic/dspace-tech/UpA9MfIqn2g/discussion) cannot be used. However, this too is probably more of a documentation issue (cf. https://jira.duraspace.org/browse/DS-2588?focusedCommentId=58931&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-58931 ) than a real problem since the lifter functionality can easily be reintroduced in a local fork.

minurmin committed:
DS-2862 Remove the call to legacy lifter.liftEmbargo, show legacy embargo check warnings only if the policy has no date

Even if the legacy embargo is used, lifting should be based on policies with dates so liftEmbargo call is not needed after all. If the policy has a date it is assumed to determine the embargo, even if the corresponding metadata field is set differently.

minurmin commented:
Based on earlier discussion the liftEmbargo line is commented again - and I have tried to clarify the reason why it remains commented. The rationale for simple null checks before showing the anonymous warnings:
* If the policy has a start date in the future, the embargo is in place so no warning is needed
* If the policy has a start date in the past, I assume that the policy date "overrides" the date in the metadata field (i.e. embargo date has been changed after the initial submission but this is reflected only in the policy) so no warning is needed here either since the item has become visible already.

terrywbrady commented:
@minurmin , your summary of the changes to the comment field make sense to me.  Thanks

tdonohue reviewed (approved):
👍 Looks good to me too. This small PR should also be applied to `master` once merged.

As a sidenote, today I updated our DSpace 6 Embargo documentation to better describe how embargoes "work" (per discussion above).  The key change is that there's now two sections describing how to "configure Embargo in Submission UI" and how to "Create embargoes via metadata". I also flagged all "lifter" functionality as deprecated / not necessary. See updated docs at: https://wiki.duraspace.org/display/DSDOC6x/Embargo

kshepherd commented:
Two +1s and changes look good to me, merging and porting to master (#2047)
